### PR TITLE
openfoam-org: fix for being able to build manually checked out repository

### PR DESCRIPTION
**Pull request by albestro**
I'm not an expert of the package and I see it is an old version, but I'm proposing this add since I addressed few requests by users asking for this specific old version.

I added this as new versions in spack because the package has few complex workaround for different versions on different repositories, and I found this to be easiest way to get the result.

### CODE REVIEW

**Comment by haampie** [comment]
I think this is fine, installing `@5` will pick `5.0` still since `x < 0` in spack ;p

**Comment by albestro** [comment]
> I think this is fine, installing `@5` will pick `5.0` still since `x < 0` in spack ;p

I don't know if having `5.x < 5.0` is a good thing in general 😅 but I wouldn't care much for this use-case.

EDIT: probably it is more a problem on how versioning is done in this specific package.

**Comment by kjrstory** [comment]
I have difficulty understanding this pull request. The actual download link for version 5.0 is the same as the GitHub address you changed it to: https://github.com/OpenFOAM/OpenFOAM-5.x/archive/version-5.0.tar.gz. However, this is the release version, and you are using the master branch version. The release version was on Jul 26, 2017, while the final commit on the master branch was on Jun 6, 2018. Are you sure a new submission is necessary due to these differences? If you need a patch for a specific release version, it seems more appropriate to use a patch rather than downloading from the GitHub branch.

**Comment by kjrstory** [comment]
Refer to this website. 

https://dl.openfoam.org/source/

While version 5.0 was sometimes referred to as 5.x, it was generally just colloquially referred to as version 5. In other words, mentioning 5.x doesn't specifically point to a particular release version. It's important to be cautious about any versions other than the official major and patch versions listed on the site. Currently, only official numbering versions are being registered in the specifications. To avoid confusion, I'm also considering changing the website address from the GitHub link to the address of this site.

**Comment by albestro** [comment]
### TL;DR
+ my proposed change about adding 5.x as listed version is a bad idea
+ there are users interested in "newer" 5.x versions
+ spack currently allows to use just 5.0 (this package does not allow to use `git.<branch>` spack feature)
+ adding to the list version `5.20171030` might be a good idea
+ there is a quite easy fix for making any manual checkout to work (so that users can patch or use any branch they want), and it will also be required in case we opt for listing `5.20171030` as "official" version.



Thanks for your message @kjrstory, it makes perfectly sense and I agree it is not a nice thing to have `5.x` as version. As already stated, it was a workaround. So, let's try to find a way out.

Let me just try to depict the situation, updated also with what I learnt from you:

### Last OpenFOAM-5 versions
- on https://dl.openfoam.org
	- `5.0`
	- `5.20171030`
- on https://github.com/OpenFOAM/OpenFOAM-5.x
	- `5.0` (the one listed in the github repo)
	- `20171030` (probably the same as the last one in dl.openfoam.org)
	- `master` (~7 months and 30 commits later https://github.com/OpenFOAM/OpenFOAM-5.x/compare/20171030...master)

The users requested last master version (I don't know the technical reasons, I'm not an OpenFOAM user as said), probably because they were following this (unofficial, but useful for them) [wiki](https://openfoamwiki.net/index.php/Installation/Linux/OpenFOAM-5.x/Ubuntu), where they actually clone the repo ending up using `master`.

Via spack it is not possible to use anything else than the versions listed (that's why I thought about that ugly workaround). Indeed, since `git` url is unified for the package and cannot be customized depending on the version, it is not possible to make use of git branches using the spec syntax `openfoam-org@git.branch=5.0`.

From your message and the overview above about versions, I would say that instead of `5.x` we could add to the list `5.20171030`, right? This might be good enough for some users (i.e. if they don't need features/fixes from any of the 30 commits that have not been released).

I think we might agree on this latter point. Anyway, I think it would not work as is (see next paragraph).

### `etc/bashrc` misalignment due to WM_PROJECT_VERSION and spack version

Indeed, one of my tries to workaround the problem was to checkout manually the code and use `spack dev-build` like this

```console
git clone https://github.com/OpenFOAM/OpenFOAM-5.x
cd OpenFOAM-5.x
spack dev-build openfoam-org@5.0
```

But unfortunately, it does not work. After your message I started looking into it more thoroughly, and I think that the problem is how `target` is computed and I would say that the solution could be found in that comment (grepping value from /etc/bashrc).

https://github.com/spack/spack/blob/2f24aeb7f690a7cc73996479b0028a273e91d9df/var/spack/repos/builtin/packages/openfoam-org/package.py#L239-L240

Indeed, build scripts like `etc/bashrc` are based on the value of `WM_PROJECT_DIR`

https://github.com/OpenFOAM/OpenFOAM-5.x/blob/20171030/etc/bashrc#L114

which is computed using `WM_PROJECT` and `WM_PROJECT_VERSION`

https://github.com/OpenFOAM/OpenFOAM-5.x/blob/20171030/etc/bashrc#L34-L35

which are hard-coded values not changed by the `patch` phase in spack (differently from `WM_PROJECT_INST_DIR` and `FOAM_INSTALL_DIR` which are set correctly).

In particular, in OpenFOAM-5.20171030 `WM_PROJECT_VERSION=5.x`, so it would not match with what `target` is set to since it uses spack version (i.e. `5.20171030` or whatever we will use as version number), and `etc/bashrc` will not find files where it expects them to be.

### EXTRA

(more a self-reminder in case we go this way)
Probably also patches for 5.0 had to be updated accordingly

https://github.com/spack/spack/blob/2f24aeb7f690a7cc73996479b0028a273e91d9df/var/spack/repos/builtin/packages/openfoam-org/package.py#L123

Not sure about `20171030`, but in `master` I'm pretty sure the patch has to be removed/changed.

**Comment by kjrstory** [comment]
@albestro
I apologize for the delay in my response due to consideration of your opinion. First, let me summarize the additional information.

1. Different flavor of OpenFOAM (ESI OpenCFD) has not only major versions but also patch versions. For instance, both major version 2212 and patch version 2212_230612 are registered.
2. If the versions are the same, the files on http://dl.openfoam.org/source/ and GitHub are identical. More precisely, the link on http://dl.openfoam.org/source/ redirects to the GitHub files.


I acknowledge the need for patch versions, as do the users. My concern is whether I should include not only patch versions for version 5 but also for versions 6 to 10, considering that there are many patch versions for each major version. While it's possible to selectively include specific patches, the challenge lies in how to make those selections. My current thought is to include patch versions for future updates from now on, while considering inclusion of past patch versions only upon request. These patch versions would be limited to officially tagged releases on GitHub (same as http://dl.openfoam.org/source/ site)

I intend to modify the package files to enable installation even for patch versions. This is simply an issue related to file paths. Given that the patch format is [MajorNumber]_[Date], I believe it shouldn't be difficult to handle. I will probably create a separate pull request for that. I will include 5.20171030 in that PR.

Users may desire versions other than patch releases. If it's not an official release (even if it's a GitHub commit), it should be considered as a user-specific version. OpenFOAM already sees users frequently making arbitrary modifications to solvers files. 
I believe that should be considered when adding site-specific packages in the Spack.  In the case of in-house code, there are instances where custom packages are added at each site. While creating packages may not be an easy task, but OpenFOAM has many user-specific versions, indicating a frequent need. If the opportunity arises, I will try my best to share the creation of package files for custom OpenFoam packages.

- Extra
Unlike the previous discussions about versions, you also mentioned the patch files. Frankly, these patch files are quite outdated, and the reasons for their necessity are unclear. A comprehensive review of all patch files is needed, not just for version 5.0. I will review the 50-etc. patch files in my PR, at least.

**Comment by albestro** [comment]
Thanks for taking the time to reply @kjrstory!

I understand your position about versions, and I agree it is better if you personally deal with it. On the other hand, in order to solve the issue I'm facing, I think I might have a proposal which looks like a win-win situation.

We can move the target of this PR to being able to `spack dev-build` a manually checked out repo of `OpenFOAM-5.x`, which currently is not possible as described in my previous comment. In this way users will be free to build the manually checked out repo (with possibly some minor changes), without adding any "tailored" version to the package.

Making this possible means:
- (required) change `rename_source()` so that it will use `WM_PROJECT_VERSION` from `eth/bashrc` and not the spack version
- (optional) restrict `patch("50-etc.patch", when="@5.0:5.9")`, e.g. apply it just for 5.0

About the latter point: the user could live even without that change, by asking to build `openfoam-org@5.10` just to workaround the range in the patch directive. IMHO it does not make much sense, and I would change it to something more strict, considering that on `master` (as I said not tested yet with `5.20171030`, but I can give it a try) that patch makes the build system fail (without it builds).

Looking forward to know what you think about this, I will start pushing the former change so that you can evaluate it. 😉

**Comment by albestro** [comment]
Just pushed the required change of the proposal and edited the title of this PR accordingly.

In case you agree, let me know (if and) how would you prefer to change this

https://github.com/spack/spack/blob/6f353a0f0f04bfefe0e901ed8428a0d4aaaf49e8/var/spack/repos/builtin/packages/openfoam-org/package.py#L123

**Comment by albestro** [comment]
ping @kjrstory

Let me know if you agree and how would you prefer to address the "patch problem" (just apply patch to 5.0 instead of the catch-all range which looks wrong?).

As soon as you let me know I'll address your comments so that we can eventually merge this PR. Thanks 😉

**Comment by kjrstory** [comment]
I agree with the grep bashrc part for now. 

The patch, as seen in #6686, seems to be for compatibility with shells other than bash, such as dash or ksh. In fact, it appears that this patch is necessary not only for version 5.0 but also for other versions. However, testing is required to confirm this. Currently, using the following should cover the patch:

```python
patch('50-etc.patch', when='@5')
```
This way, the patch will be applied to all versions of 5.

**Review by kjrstory — APPROVED** [review]

**Comment by albestro** [comment]
Thanks for pointing that out @kjrstory!

I had a quick look to the PR you mentioned and also at the repo and I'm pretty sure that the error I'm getting when that patch is applied, is because in versions later then 5.0 it is already applied!

Indeed, also in `5.20171030` there's a commit https://github.com/OpenFOAM/OpenFOAM-5.x/commit/c66fba323c348012c2d4a75b9091406e1b3bcc31 very similar (almost identical) to what the patch `50-etc.patch` does. So the patch should not be applied to other than `5.0` otherwise it errors

References:
- https://github.com/OpenFOAM/OpenFOAM-5.x/commits/20171030/etc/bashrc)
- https://github.com/OpenFOAM/OpenFOAM-5.x/blob/197d9d3bf20ac0641a8fe4f57be6a6325ca834b1/etc/bashrc#L45-L47

I have to change that, otherwise nobody will be able to build with spack anything later than 5.0, even if checked out manually (which is the target of this PR).

Waiting for your feedback 😉

**Review by kjrstory — APPROVED** [review]

**Comment by kjrstory** [comment]
@albestro 
Thank you very much for pointing out that commit. Now, it seems appropriate to ping someone with merge permissions.

**Review by alalazo — APPROVED** [review]
